### ISSUE TITLE
pip.installed with upgrade: true cause failed

### ISSUE THREAD

**Issue by esviderskii** [saltstack/salt#48223]
### Setup
sls

```
pack_install:
  pip.installed:
....
    - name: pack
    - extra_index_url: "https://user:password@privaterepo.com/simple"   # private repo
    - upgrade: True
....
```
result (-l debug)

```
[INFO    ] Running state [pack] at time 12:36:15.678383
[INFO    ] Executing state pip.installed for [pack]
[DEBUG   ] pip: Found python binary: /opt/pack/venv/bin/python
[INFO    ] Executing command [u'/opt/pack/venv/bin/python', u'-m', u'pip', u'--version'] in directory '/root'
[DEBUG   ] stdout: pip 10.0.1 from /opt/pack/venv/lib/python3.5/site-packages/pip (python 3.5)
[DEBUG   ] Installed pip version: 10.0.1
[DEBUG   ] pip: Found python binary: /opt/pack/venv/bin/python
[INFO    ] Executing command [u'/opt/pack/venv/bin/python', u'-m', u'pip', u'freeze', u'--all'] in directory '/root'
[DEBUG   ] stdout: aiohttp==1.2.0
async-timeout==2.0.1
cachetools==1.1.6
chardet==3.0.4
marshmallow==2.15.3
multidict==4.3.1
numpy==1.14.5
pack==1.4.17
pip==10.0.1
pkg-resources==0.0.0
python-dateutil==2.7.3
raven==6.2.1
setuptools==39.2.0
six==1.11.0
wheel==0.31.1
yarl==0.8.1
[DEBUG   ] pip: Found python binary: /opt/pack/venv/bin/python
[ERROR   ] An exception occurred in this state: Traceback (most recent call last):
  File "/usr/lib/python2.7/dist-packages/salt/state.py", line 1905, in call
    **cdata['kwargs'])
  File "/usr/lib/python2.7/dist-packages/salt/loader.py", line 1830, in wrapper
    return f(*args, **kwargs)
  File "/usr/lib/python2.7/dist-packages/salt/states/pip_state.py", line 726, in installed
    **kwargs)
  File "/usr/lib/python2.7/dist-packages/salt/states/pip_state.py", line 237, in _check_if_installed
    desired_version = available_versions[-1]
TypeError: 'NoneType' object has no attribute '__getitem__'

[INFO    ] Completed state [pack] at time 12:36:17.848666 (duration_in_ms=2170.282)
```

If change `upgrade: False` or added `force_reinstall: True` - run and executed without errors.

### Versions Report

salt-call version
```
Salt Version:
           Salt: 2018.3.1
 
Dependency Versions:
           cffi: Not Installed
       cherrypy: Not Installed
       dateutil: 2.4.2
      docker-py: Not Installed
          gitdb: Not Installed
      gitpython: Not Installed
          ioflo: Not Installed
         Jinja2: 2.8
        libgit2: Not Installed
        libnacl: Not Installed
       M2Crypto: Not Installed
           Mako: 1.0.3
   msgpack-pure: Not Installed
 msgpack-python: 0.4.6
   mysql-python: Not Installed
      pycparser: Not Installed
       pycrypto: 2.6.1
   pycryptodome: Not Installed
         pygit2: Not Installed
         Python: 2.7.12 (default, Dec  4 2017, 14:50:18)
   python-gnupg: 0.3.8
         PyYAML: 3.11
          PyZMQ: 15.2.0
           RAET: Not Installed
          smmap: Not Installed
        timelib: Not Installed
        Tornado: 4.2.1
            ZMQ: 4.1.4
 
System Versions:
           dist: Ubuntu 16.04 xenial
         locale: ANSI_X3.4-1968
        machine: x86_64
        release: 4.9.0-6-amd64
         system: Linux
        version: Ubuntu 16.04 xenial

```

**Comment by Ch3LL** [comment]
im having a hard time replicating this using the following state:

```
python-pip:
  pip.installed:
    - name: pudb
    - upgrade: True
```

Do you run into the issue if you run this simple state above as well?

**Comment by esviderskii** [comment]
No, тhe problem appears when there is an extra_index_url in the state

**Comment by Ch3LL** [comment]
I'm still not able to replicate this with this state:


```
pudb:
  pip.installed:
    - index_url: https://localpipserver
    - extra_index_url: https://pypi.python.org/simple
    - upgrade: True
```

```
➜  salt git:(9aada36c1c) sudo salt-call --local state.apply issues.48223 -lquiet
local:
----------
          ID: pudb
    Function: pip.installed
      Result: True
     Comment: All packages were successfully installed
     Started: 14:37:57.840334
    Duration: 4513.305 ms
     Changes:   
              ----------
              pudb==2018.1:
                  Installed

Summary for local
------------
Succeeded: 1 (changed=1)
Failed:    0
------------
Total states run:     1
Total run time:   4.513 s
➜  salt git:(9aada36c1c) sudo salt-call --local state.apply issues.48223 -lquiet
local:
----------
          ID: pudb
    Function: pip.installed
      Result: True
     Comment: Python package pudb was already installed
              All specified packages are already installed and up-to-date
     Started: 14:38:17.280962
    Duration: 1335.849 ms
     Changes:   

Summary for local
------------
Succeeded: 1
Failed:    0
------------
Total states run:     1
Total run time:   1.336 s
```

as you can see running it twice I get the expected return. Any other ideas as to why your use case is running into this and not mine?

**Comment by selfin** [comment]
Problem is reproducible only for packages available only from a custom index.
The reason for such behavior is that pip module function `list_all_versions`  does not accept `extra_index_url` param.
`extra_index_url` support should be added as previously added index_url: https://github.com/saltstack/salt/pull/45945

**Comment by rallytime** [comment]
@esviderskii This should be fixed with #48419
